### PR TITLE
Adds paren to deps for hidden extra constraint

### LINKED ISSUE #4107: Dependencies for `black[d]` are being installed by default
**Describe the bug**

Minor bug, but it seems that the extra dependencies for the `d` extra are being installed by default on 23.12.0. This can cause some issues on docker containers and such where `gcc` isn't installed, as the extra dependencies (`aiohttp`, and `frozenlist`) don't seem to have wheels.

**To Reproduce**

On a non-windows machine (I've done it on Ubuntu, and it also seems to happen on Linux github action runners) Make a new venv, source it, and install black:

```bash
python3 -m venv .venv
source .venv/bin/activate
pip install black
```

It will install requirements from the `d` extra as well; 

```
Installing collected packages: platformdirs, packaging, tomli, multidict, idna, yarl, attrs, frozenlist, async-timeout, aiosignal, aiohttp, typing-extensions, click, pathspec, mypy-extensions, black
Successfully installed aiohttp-3.9.1 aiosignal-1.3.1 async-timeout-4.0.3 attrs-23.1.0 black-23.12.0 click-8.1.7 frozenlist-1.4.0 idna-3.6 multidict-6.0.4 mypy-extensions-1.0.0 packaging-23.2 pathspec-0.12.1 platformdirs-4.1.0 tomli-2.0.1 typing-extensions-4.9.0 yarl-1.9.4
```

**Expected behavior**

It shouldn't install with the `aiohttp`, `frozenlist`, and `multidict` dependencies.

**Environment**

- Black's version: 23.12.0, pip version 20.0.2
- OS and Python version: Python 3.8.10, Ubuntu 20.04, though it also happens with github runners and the latest python-3 docker container on 22.04, see https://github.com/SuffolkLITLab/docassemble-AssemblyLine/actions/runs/7182833153/job/19560217190 

**Additional context**

First seen in https://github.com/rickstaa/action-black/issues/18

I think it has to do with: https://github.com/psf/black/commit/85b1c71a3445f32860d7b139ae4de4824f6ae102; in my pip logs, the new restrictions seem to show up as:

```
Collecting aiohttp>=3.7.4; sys_platform != "win32" or implementation_name != "pypy" and extra == "d"
```

i.e., the `extra == "d"` is ignored in favor of the sys platform not being windows, so the extras are always installed. I think a quick fix would be to wrap the extras in parenthesis. Happy to make a quick PR for that.

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
### Description

Fix #4107 by updating hatchling  ~~in a fairly hacky way though.~~

Hatch ~~will~~ used to strip out parenthesis in the extras if it thinks it doesn't need them and wouldn't add them back in. ~~:( We can, however, add a meaningless restriction like `and python_version >= '1'` to force hatch to use our parenthesis. Extremely hacky, and I would understand if we don't want this to be merged until a cleaner hatch fix is upstream. Also happy to come up with another way of essentially just having `true` in the spec, but I couldn't find one with a quick reading of https://peps.python.org/pep-0508/.~~

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?   I'm not sure it's a great idea to add a test to see that dependency was avoided, happy to add one if desired though!
      I did manually test this by installing my local checkout of the project, confirming that the dependencies for the `d` extra aren't installed:
      
```logs
$ pip install .
Processing /home/brycew/Developer/black
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: typing-extensions>=4.0.1 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (4.9.0)
Requirement already satisfied: tomli>=1.1.0 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (2.0.1)
Requirement already satisfied: mypy-extensions>=0.4.3 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (1.0.0)
Requirement already satisfied: pathspec>=0.9.0 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (0.12.1)
Requirement already satisfied: platformdirs>=2 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (4.1.0)
Requirement already satisfied: click>=8.0.0 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (8.1.7)
Requirement already satisfied: packaging>=22.0 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1763+gdad9f79.d20231212) (23.2)
Building wheels for collected packages: black
  Building wheel for black (pyproject.toml) ... done
  Created wheel for black: filename=black-0.1.dev1763+gdad9f79.d20231212-py3-none-any.whl size=194930 sha256=a511959e5af87f2bc5623744f13eba2bd3c406fd9b522e8a537c8f3d341b1a18
  Stored in directory: /tmp/pip-ephem-wheel-cache-o2gmqktu/wheels/92/d6/de/b610e33302c2b433609d1ae2283761d3ff26bd0fb805d38fe1
Successfully built black
Installing collected packages: black
  Attempting uninstall: black
    Found existing installation: black 0.1.dev1763+gdad9f79
    Uninstalling black-0.1.dev1763+gdad9f79:
      Successfully uninstalled black-0.1.dev1763+gdad9f79
Successfully installed black-0.1.dev1763+gdad9f79.d20231212
```

When run on main, this output includes `Requirement already satisfied: aiohttp>=3.7.4 in /home/brycew/Developer/.venv/lib/python3.10/site-packages (from black==0.1.dev1762+g35ce37d) (3.9.1)`, and all of it's dependencies.

- [x] Add new / update outdated documentation? (nothing changed here)

### CODE REVIEW

**Review by AlexWaygood — COMMENTED** [review]
Thanks for fixing! Now that https://github.com/pypa/hatch/releases/tag/hatchling-v1.20.0 has been released with an upstream fix, maybe a simpler solution might be to just pin to `hatchling>=1.20.0` here?

https://github.com/psf/black/blob/35ce37ded7bd8fdd3950af19e7c11f311ee7b8d8/pyproject.toml#L28

**Comment by BryceStevenWilley** [comment]
Thanks! I just saw that hatchling released the patch already, which is nice. I also changed the other pinned version of hatchling, I don't have a mac machine to test that one though.

**Review by hauntsaninja — APPROVED** [review]
Thanks for the report and for fixing this twice over!

We'll test the Mac build in CI (looks like the patch still applies, which is main reason for that pin)

File: CHANGES.md [hunk 1]
```diff
@@ -22,6 +22,8 @@
 
 <!-- Changes to how Black is packaged, such as dependency requirements -->
 
+- Fixed a bug that included dependencies from the `d` extra by default (#4108, #4107).
```
Comment (hauntsaninja) [inline CHANGES.md:25]:
```suggestion
- Fixed a bug that included dependencies from the `d` extra by default (#4108)
```

**Comment by sigma67** [comment]
Any plans for a patch release including this fix?

**Pull request merged**